### PR TITLE
[14.0][MIG] product_stock_state

### PR DESCRIPTION
**Pull request by watthanun**
standard migration

### CODE REVIEW

**Comment by kevinkhao** [comment]
LGTM
khapkun

**Review by kevinkhao — APPROVED** [review]

File: product_stock_state/models/product_product.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @api.depends(
+        "qty_available",
+        "incoming_qty",
+        "stock_state_threshold",
+        "company_id.stock_state_threshold",
+    )
+    def _compute_stock_state(self):
+        for product in self:
+            qty_available = product._get_qty_available_for_stock_state()
+            precision = self.env["decimal.precision"].precision_get("Stock Threshold")
```
Comment (rousseldenis) [inline product_stock_state/models/product_product.py]: Please put this outside of the loop

File: product_stock_state/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# @author Sylvain LE GAL <https://twitter.com/legalsylvain>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    stock_state_threshold = fields.Float(default=10, digits="Stock Threshold")
```
Comment (rousseldenis) [inline product_stock_state/models/res_company.py:11]: Shouldn't this use decimal precision ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

File: product_stock_state/models/product_category.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        " company level.",
+        digits="Stock Threshold",
+    )
+    manual_stock_state_threshold = fields.Float(digits="Stock Threshold")
+
+    @api.depends("parent_id.stock_state_threshold", "manual_stock_state_threshold")
+    def _compute_stock_state_threshold(self):
+        for rec in self:
+            rec.stock_state_threshold = (
+                rec.manual_stock_state_threshold or rec.parent_id.stock_state_threshold
```
Comment (petrus-v) [inline product_stock_state/models/product_category.py:30]: I'm wondering what's happen if rec.manual_stock_state_threshold is falsy (=0.0) and there are no parent_id set on the record ?

File: product_stock_state/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def _get_qty_available_for_stock_state(self):
+        """
+        This method can be overridden to provide the available qty.
+        In some cases you could prefer to use the qty_available - outgoing_qty
+        to take into account products reserved
+        """
+        self.ensure_one()
+        return self.qty_available
+
+    def _stock_state_check_in_stock(self, qty, precision):
```
Comment (petrus-v) [inline product_stock_state/models/product_product.py:41]: I'm wondering where and why those several private methods are used for ?

**Review by ferranmateu — APPROVED** [review]
LGTM!

**Comment by rousseldenis** [comment]
@watthanun Could you attend comments?

**Comment by rousseldenis** [comment]
@watthanun @kittiu

File: product_stock_state/models/product_product.py (hunk 1 above)
Comment (kittiu) [reply product_stock_state/models/product_product.py]: Fixed

File: product_stock_state/models/res_company.py (hunk 2 above)
Comment (kittiu) [reply product_stock_state/models/res_company.py:11]: It is, isn't it ?  `digits="Stock Threshold"` :smile_cat:

File: product_stock_state/models/product_product.py (hunk 4 above)
Comment (kittiu) [reply product_stock_state/models/product_product.py:41]: Removed.

**Comment by kittiu** [comment]
> @watthanun @kittiu

@rousseldenis sorry for delayed. I am much occupied these days. Please feel free to nag me, if need. :)

File: product_stock_state/models/product_category.py (hunk 3 above)
Comment (kittiu) [reply product_stock_state/models/product_category.py:30]: :thinking: may be zero.

File: product_stock_state/models/product_product.py (hunk 4 above)
Comment (petrus-v) [reply product_stock_state/models/product_product.py:41]: @kittiu thanks for that one, I think it's the same for `_stock_state_check_in_limited_stock`, `_stock_state_check_resupplying`, and even more `_stock_state_check_out_of_stock`.

Comment (kittiu) [reply product_stock_state/models/product_product.py:41]: Thanks!

Comment (petrus-v) [reply product_stock_state/models/product_product.py:41]: :+1:

**Review by petrus-v — APPROVED** [review]

File: product_stock_state/models/product_product.py (hunk 4 above)
Comment (sebastienbeau) [reply product_stock_state/models/product_product.py:41]: @kittiu @petrus-v There is a misundertood here, we should not remove the method. The module is broken now.
Indeed this method are called here : https://github.com/OCA/product-attribute/pull/715/commits/cf2c5a1bd6e9b7fc778466fb9941c6d5a3f5e1ee#diff-a0f343e91cae7060fd8ed792256dc90b6b263767beab71becc4e51be5954002bR55

Thanks for your time for the migration

**Review by sebastienbeau — CHANGES_REQUESTED** [review]
Module is broken see my comment: https://github.com/OCA/product-attribute/pull/715#discussion_r609636094

File: product_stock_state/models/product_product.py (hunk 4 above)
Comment (petrus-v) [reply product_stock_state/models/product_product.py:41]: ok I missed that code sorry, but, but... there were not test to prevents removing those methods ;-)

Comment (petrus-v) [reply product_stock_state/models/product_product.py:41]: hum...  CI is broken probably some of them were covered sorry

**Comment by kittiu** [comment]
Sorry guys, I was much caught up these days. Fixed now.

**Review by rousseldenis — APPROVED** [review]

**Comment by rousseldenis** [comment]
@sebastienbeau

**Review by sebastienbeau — APPROVED** [review]

**Pull request merged**